### ISSUE TITLE
using gofmt and go vet in sh_test rules

### ISSUE THREAD

**Issue by ixdy** [bazelbuild/rules_go#511]
We have a few lint-style checks for `gofmt` and `go vet` which are implemented as `sh_test`s. We had a `data` dependency on `@io_bazel_rules_go_toolchain//:toolchain`, which brought `external/io_bazel_rules_go_toolchain/bin` into the runfiles for the test, but with the recent refactoring, the `:toolchain` target no longer exists.

I tried updating this to `:go_toolchain`, but that doesn't seem to include any of the Go toolchain in the test runfiles anymore.

Is there a way we can depend on the Go toolchain from outside the `rules_go` repository? I've tried searching through the changes, but I'm having trouble figuring out if this is supposed to be possible.

cc @spxtr

**Comment by jayconrod** [comment]
I don't think this was supposed to be possible, but it seems like a good use case. Maybe we could expose some stable aliases to tools like `gofmt`. @ianthehat WDYT?

**Comment by ianthehat** [comment]
This is possible, but it depends exactly what you want.

There are now two major parts, the go sdk (per platform and version), and the go toolchain (for various combinations of options and cross compiles and underlying sdk).

Toolchain selection logic is currently a temporary hack while we wait for true toolchain selection based on platform constraints to appear in Bazel, which is why some of the rules are a bit obtuse right now, it also means that some of the ways of getting to the toolchain and sdk that I might suggest will also be broken again as soon as that feature arrives.

It sounds like what you really want is the sdk not the toolchain, but without the toolchain there is no way to know which sdk is in use, so it is probably best to go through the toolchain to get to the sdk.
For now, you use an attribute
```"_go_toolchain": attr.label(default = Label("@io_bazel_rules_go_toolchain//:go_toolchain")),```
to bind your dependancy on the toolchain selector, but that will give you something that is not directly useful in a data dependancy. 
Also, we no longer capture the entire bin directory from the sdk, only the go binary and the pkg/tools binaries, so if you are using gofmt,  we would have to add that as well.

Maybe the right thing to do here is for us to talk about how to provide this functionality directly, it seems like something most users of the rules_go might want anyway, and their might be nicer ways to do it. Could you share your current approach?

**Comment by ixdy** [comment]
Sure. We use `gofmt` and `go vet` as lint checks on our repositories. We have pretty simple shell wrappers around them ([`verify-gofmt.sh`](https://github.com/kubernetes/test-infra/blob/master/verify/verify-gofmt.sh), [`verify-govet.sh`](https://github.com/kubernetes/test-infra/blob/master/verify/verify-govet.sh)) and then turn them into Bazel tests by depending on all sources in the tree ([`BUILD`](https://github.com/kubernetes/test-infra/blob/25db62f5a275343e395f5d2032f112e880be4d6f/verify/BUILD#L11-L27)).

It's a little strange, since this isn't exactly something Bazel is designed to do, but it works well for us.

We want to follow the Go version used to build the code, since formatting/vet rules tend to change between Go versions. Thus using the sdk/toolchain from `go_rules` makes sense for us.

If we could depend on the selected `gofmt` and `go` binaries directly, that would solve the problem for us. (I'm not familiar with all of the toolchain changes, so I don't now how that plays in to everything.)

**Comment by ixdy** [comment]
Another possibility is exposing `go_vet` and `go_fmt` test rules. I'm not sure how one would best manage the source dependencies, though: users could do something like our `all-srcs` rules, or they could be automanaged by gazelle?

I think @spxtr and @mikedanese have other ideas, too.

**Comment by ash2k** [comment]
In addition to `go_fmt` test rule it'd be useful to have a rule that formats the source files.  It should probably format only files/packages that were modified. Or this is possible already (without manually editing each `BUILD` file)? I'm very new to bazel/rules_go.

**Comment by ianthehat** [comment]
The thing preventing me from getting this fully done right now is the inability to know what directory bazel run was invoked from.
I have started a discussion about this, as soon as we have that I can add the rest of the functionality trivially (based on the work I already did for the gazelle rule)
Only changed files or parts of files is much harder because it requires knowledge of what you mean by changed, which needs understanding of all the possible vcs systems, probably better and easier to always apply it to the whole file/files/targets specified.

**Comment by jayconrod** [comment]
@drigz Right, there's no longer anything blocking us from running a tool invoked by `bazel run` that writes files in the source directory.

**Comment by ash2k** [comment]
Related to this issue - if anyone is interested, we have a rule to run [goimports](https://github.com/atlassian/bazel-tools/tree/master/goimports). I'm interested in any feedback.

**Comment by charlesoconor** [comment]
If `gofmt` was exposed from the rules here it could be integrated into a tool like https://github.com/thundergolfer/bazel-linting-system which currently relies on go to be installed on the system which isn't very hermtic.

**Comment by SlyMarbo** [comment]
If there's any interest in adding `gofmt` checking to rules_go, I'd be happy to have a go at implementing it. I've just written a naive implementation for another project [here](https://github.com/ProjectSerenity/firefly/commit/f6ee59bfadea4bedac2fe8d724a79d3412a63873). I've used the `gofmt` binary from the Go SDK produced by rules_go by referencing `@go_sdk//:bin/gofmt`. I then use a simple builder tool to invoke `gofmt -s -l` on a set of source files, extracted from one or more `GoSource` providers. Using the rule looks like this:

```bazel
go_library(
    name = "lib",
    srcs = [...],
)

gofmt(
    name = "gofmt",
    embed = [":lib"],
)
```

If the code is formatted, the 'build' succeeds, producing an empty file. If the code is not formatted, then the build fails, printing the list of filenames that require formatting.

One thing to consider is whether the rule should behave as a build or as a test. Running as a build (as described above) doesn't feel quite right logically, but it's simple to implement. Running as a test makes more sense, but it's a little more fiddly to implement and has the extra downside of making things like `bazel test //...` more noisy, as most Go packages will now have an extra test target. Not a big problem, but something to consider.

**Comment by derekperkins** [comment]
This is somewhat out of scope, but fwiw, we haven't run those as individual tools in years, they're just a part of a single golangci-lint configuration. Getting that to run would require some more work, but then any tool moving forward would just have to integrate once with golangci-lint, and bazel would come for free. There was some progress made a while back, but never pushed over the finish line.
- https://github.com/bazelbuild/rules_go/issues/2695
- https://github.com/golangci/golangci-lint/issues/1473

**Comment by Zemnmez** [comment]
I've been porting my repo over to `MODULE.bazel`, and in [the walkthrough] it states about using the SDK:

[the walkthrough]: https://github.com/bazelbuild/rules_go/blob/93f29a542553f50787233cac7d9006ce300d4079/docs/go/core/bzlmod.md#using-a-go-sdk

> If you have a use case that would require this, please explain it in an issue.

I'm trying to do the above. I'm using `aspect_rules_lint`, which requires a `gofmt` binary I previously got from `go_sdk`. I would appreciate a supported way to run the `stdlib/cmd` tools in the Go SDK.
